cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/Job.java

```diff
@@ -40,27 +43,27 @@
     /**
      * The Job is awaiting fulfillment of its Trigger.
      */
-    PENDING_TRIGGER,
+    PENDING_LAUNCH(Collections.<State>emptySet()),
     /**
      * The Job is awaiting satisfaction of all of its constraints.
      */
-    PENDING_CONSTRAINT,
+    PENDING_CONSTRAINT(EnumSet.of(PENDING_LAUNCH)),
     /**
      * The Job is awaiting launch.
      */
-    PENDING_LAUNCH;
+    PENDING_TRIGGER(EnumSet.of(PENDING_CONSTRAINT));
 
-    private static final Map<State, State> possibleTransitions = new HashMap<>();
-    static {
-      possibleTransitions.put(PENDING_TRIGGER, PENDING_CONSTRAINT);
-      possibleTransitions.put(PENDING_CONSTRAINT, PENDING_LAUNCH);
+    private final Set<State> allowedNextState;
+
+    State(Set<State> allowedNextState) {
+      this.allowedNextState = allowedNextState;
     }
```

Comment:
How that? That depends on the algorithm that transitions these states. If the constraint checker implements an optimization to transition from PENDING_TRIGGER directly to PENDING_LAUNCH, if there are no run constraints, then that is perfectly valid. The state itsefl cannot know what transitions are possible.